### ISSUE TITLE
[Raft] Peer cannot rejoin after being briefly disconnected

### ISSUE THREAD

**Issue by NatoBoram** [ipfs/ipfs-cluster#718]
#### Pre-check

* [x] This is not a IPFS Cluster website content issue ([file those here](https://github.com/ipfs/ipfs-cluster-website/issues))
* [x] I read the [troubleshooting section](https://cluster.ipfs.io/documentation/troubleshooting/) of the website and it did not help
* [x] I searched for similar issues in the repo without luck
* [x] All my peers are running the same cluster version
* [x] All my peers are configured using the same cluster `secret`

#### Basic information

* [x] Version information (mark as appropiate):
  * [x] Master
  * [ ] Release candidate for next version
  * [ ] Latest stable version
  * [ ] An older version I should not be using
* [x] Type (mark as appropiate):
  * [x] Bug
  * [ ] Feature request
  * [ ] Enhancement
* [x] Operating system (mark as appropiate):
  * [x] Linux
  * [ ] macOS
  * [ ] Windows
  * [ ] Other: `which?`
* [x] Installation method (mark as appropiate):
  * [ ] Binaries from dist.ipfs.io
  * [x] Built from sources
  * [ ] Docker
  * [ ] Snap
  * [ ] Other: `which?`

#### Description

When a peer is disconnected, it cannot re-join the existing cluster.

Situation : 

* I have 3 peers : `lineageos-on-ipfs`, `nato-hp`, and `nato-helion`.
* `nato-helion` and `nato-hp` are in the same network.
  * They have leave-on-shutdown enabled.
  * The router reboots every night
* `nato-helion` goes down regularly.

After a reboot of `nato-helion`, `ipfs-cluster-ctl status` on `lineageos-on-ipfs` will indicate that a peer is missing.

```
QmaimL3hrBQ7ocLcokm7X7CnNA1e2t4aXH3t2mGHfAsMYj :
    > hp-server       : PINNED | 2019-03-14T02:24:25.078904428Z
    > QmTPuNCYQJTdsJ8xUZMGeZ9ReecMbsguaEqyk26s8TFUDV : CLUSTER_ERROR: dial attempt failed: QmUxeeoYLoxEQq37pPo2BiFPjcYvJLqeHBRMfL6YtWqtUA --> QmTPuNCYQJTdsJ8xUZMGeZ9ReecMbsguaEqyk26s8TFUDV dial attempt failed: dial tcp4 0.0.0.0:9096->209.15.82.76:9096: i/o timeout | 2019-03-14T04:31:15.3377328Z
    > lineageos-on-ipfs : REMOTE | 2019-03-14T02:24:22.077327709Z
```

However, I enabled IPFS-Cluster as a service on all machines. Upon reboot, they should automatically reconnect to the cluster.

```service
[Unit]
Description=Inter-Planetary File System Cluster

[Service]
ExecStart=/home/nato/go/bin/ipfs-cluster-service daemon --upgrade --bootstrap /ip4/159.89.116.13/tcp/9096/ipfs/QmUxeeoYLoxEQq37pPo2BiFPjcYvJLqeHBRMfL6YtWqtUA
Restart=on-failure

[Install]
WantedBy=default.target
```

If I start it manually, here's the errors : 

```
00:35:11.262  INFO    service: Initializing. For verbose output run with "-l debug". Please wait... daemon.go:45
00:35:11.263 WARNI     config: metrics section is empty, generating default config.go:370
00:35:11.263 WARNI     config: tracing section is empty, generating default config.go:370
00:35:11.264 WARNI     config: tracing section is empty, generating default config.go:370
00:35:11.264 WARNI     config: metrics section is empty, generating default config.go:370
00:35:11.264  INFO  consensus: cleaning empty Raft data folder (/home/nato/.ipfs-cluster/raft) raft.go:657
00:35:11.264 WARNI    service: the /home/nato/.ipfs-cluster/raft folder has been rotated.  Next start will use an empty state state.go:184
00:35:11.474  INFO    cluster: IPFS Cluster v0.10.0+git508131081124bc8289b762ffb7e0e0572798e9be listening on:
        /p2p-circuit/ipfs/QmTPuNCYQJTdsJ8xUZMGeZ9ReecMbsguaEqyk26s8TFUDV
        /ip4/127.0.0.1/tcp/9096/ipfs/QmTPuNCYQJTdsJ8xUZMGeZ9ReecMbsguaEqyk26s8TFUDV
        /ip4/192.168.1.2/tcp/9096/ipfs/QmTPuNCYQJTdsJ8xUZMGeZ9ReecMbsguaEqyk26s8TFUDV

 cluster.go:114
00:35:11.475  INFO  ipfsproxy: IPFS Proxy: /ip4/127.0.0.1/tcp/9095 -> /ip4/127.0.0.1/tcp/5001 ipfsproxy.go:273
00:35:11.475  INFO  consensus: peer is ready to join a cluster raft.go:205
00:35:11.475  INFO    service: Bootstrapping to /ip4/159.89.116.13/tcp/9096/ipfs/QmUxeeoYLoxEQq37pPo2BiFPjcYvJLqeHBRMfL6YtWqtUA daemon.go:172
00:35:11.475  INFO    restapi: REST API (HTTP): /ip4/127.0.0.1/tcp/9094 restapi.go:456
00:35:11.475  INFO    restapi: REST API (libp2p-http): ENABLED. Listening on:
        /p2p-circuit/ipfs/QmTPuNCYQJTdsJ8xUZMGeZ9ReecMbsguaEqyk26s8TFUDV
        /ip4/127.0.0.1/tcp/9096/ipfs/QmTPuNCYQJTdsJ8xUZMGeZ9ReecMbsguaEqyk26s8TFUDV
        /ip4/192.168.1.2/tcp/9096/ipfs/QmTPuNCYQJTdsJ8xUZMGeZ9ReecMbsguaEqyk26s8TFUDV

 restapi.go:473
00:35:22.590  INFO  consensus: Current Raft Leader: QmPp9hcP4ZzbkawU2nZ41kVQwHCZvdwenagQ6qzXWsJFnZ raft.go:280

00:35:22.599 ERROR   ipfshttp: error posting to IPFS: Post http://127.0.0.1:5001/api/v0/pin/ls?arg=QmWCVHhoMTBrbJZ3Sbdp7FSUNh6dvE9Xtu4f3dRBC7CGMD&type=recursive: context canceled ipfshttp.go:431
00:35:22.599 ERROR  p2p-gorpc: Post http://127.0.0.1:5001/api/v0/pin/ls?arg=QmWCVHhoMTBrbJZ3Sbdp7FSUNh6dvE9Xtu4f3dRBC7CGMD&type=recursive: context canceled call.go:63

00:35:23.263  INFO    cluster: Cluster Peers (without including ourselves): cluster.go:445
00:35:23.263  INFO    cluster:     - QmPp9hcP4ZzbkawU2nZ41kVQwHCZvdwenagQ6qzXWsJFnZ cluster.go:452
00:35:23.263  INFO    cluster:     - QmUxeeoYLoxEQq37pPo2BiFPjcYvJLqeHBRMfL6YtWqtUA cluster.go:452
00:35:23.263  INFO    cluster: ** IPFS Cluster is READY ** cluster.go:460
00:35:23.263  INFO    cluster: QmTPuNCYQJTdsJ8xUZMGeZ9ReecMbsguaEqyk26s8TFUDV: joined QmUxeeoYLoxEQq37pPo2BiFPjcYvJLqeHBRMfL6YtWqtUA's cluster cluster.go:760
00:35:23.311 ERROR   ipfshttp: error posting to IPFS: Post http://127.0.0.1:5001/api/v0/pin/ls?arg=QmSbyFDGDrHtr2LKFX2fuSF3cChD54ZvphJKhfsh1SY7ZL&type=recursive: context canceled ipfshttp.go:431

00:35:23.311 ERROR  p2p-gorpc: Post http://127.0.0.1:5001/api/v0/pin/ls?arg=QmSbyFDGDrHtr2LKFX2fuSF3cChD54ZvphJKhfsh1SY7ZL&type=recursive: context canceled call.go:63
00:35:23.311 ERROR   ipfshttp: error posting to IPFS: Post http://127.0.0.1:5001/api/v0/pin/ls?arg=QmYobGuRsDfiMpVgfgwNzGJW12QD4qT5KqK3SzZABrwkJB&type=recursive: context canceled ipfshttp.go:431

00:35:28.325  INFO    cluster: QmTPuNCYQJTdsJ8xUZMGeZ9ReecMbsguaEqyk26s8TFUDV: removed from raft. Initiating shutdown cluster.go:357
00:35:28.325  INFO    cluster: shutting down Cluster cluster.go:483
00:35:28.325  INFO  consensus: stopping Consensus component consensus.go:185
00:35:28.326 ERROR       raft: NOTICE: Some RAFT log messages repeat and will only be logged once logging.go:105
00:35:28.326 ERROR       raft: Failed to take snapshot: cannot take snapshot now, wait until the configuration entry at 2173 has been applied (have applied 1534) logging.go:105
00:35:28.326 WARNI  consensus: retrying to snapshot (1/5)... raft.go:425
00:35:28.326 WARNI  consensus: retrying to snapshot (2/5)... raft.go:425
00:35:28.326 WARNI  consensus: retrying to snapshot (3/5)... raft.go:425
00:35:28.326 WARNI  consensus: retrying to snapshot (4/5)... raft.go:425
00:35:28.326 WARNI  consensus: retrying to snapshot (5/5)... raft.go:425
00:35:28.326  INFO  consensus: cleaning empty Raft data folder (/home/nato/.ipfs-cluster/raft) raft.go:657
00:35:28.326  INFO    monitor: stopping Monitor pubsubmon.go:162
00:35:28.326  INFO    restapi: stopping Cluster API restapi.go:494
00:35:28.327  INFO  ipfsproxy: stopping IPFS Proxy ipfsproxy.go:248
00:35:28.327  INFO   ipfshttp: stopping IPFS Connector ipfshttp.go:207
00:35:28.327  INFO pintracker: stopping MapPinTracker maptracker.go:124
00:35:28.327 ERROR  p2p-gorpc: context canceled call.go:63
00:35:28.327 ERROR  p2p-gorpc: context canceled call.go:63
00:35:28.327 ERROR  p2p-gorpc: context canceled call.go:63
00:35:28.327 ERROR  p2p-gorpc: context canceled call.go:63
```

Now, I'll try to remove this peer from the cluster elsewhere, cleanup the state, and rejoin the cluster.

```bash
ipfs-cluster-ctl peers rm QmTPuNCYQJTdsJ8xUZMGeZ9ReecMbsguaEqyk26s8TFUDV
```

That seemed to work, since it disappeared from the status.

```
QmbbYWqZ2TqmZPxP41zNBnmXAMXQeURB8ESSF2wYbuYmyU :
    > hp-server       : PINNING | 2019-03-13T19:12:40.455085793Z
    > lineageos-on-ipfs : REMOTE | 2019-03-14T00:31:31.810050465Z
```

Now let's try to join again, from fresh.

```bash
ipfs-cluster-service state cleanup
ipfs-cluster-service daemon --upgrade --bootstrap /ip4/159.89.116.13/tcp/9096/ipfs/QmUxeeoYLoxEQq37pPo2BiFPjcYvJLqeHBRMfL6YtWqtUA
```

```
The peer's state will be removed from the load path.  Existing pins may be lost. Continue? [y/n]: y
00:40:05.039 WARNI     config: metrics section is empty, generating default config.go:370
00:40:05.039 WARNI     config: tracing section is empty, generating default config.go:370
00:40:05.040  INFO  consensus: cleaning empty Raft data folder (/home/nato/.ipfs-cluster/raft) raft.go:657
00:40:05.040 WARNI    service: the /home/nato/.ipfs-cluster/raft folder has been rotated.  Next start will use an empty state state.go:184
```

```
00:41:04.538  INFO    service: Initializing. For verbose output run with "-l debug". Please wait... daemon.go:45
00:41:04.539 WARNI     config: metrics section is empty, generating default config.go:370
00:41:04.539 WARNI     config: tracing section is empty, generating default config.go:370
00:41:04.540 WARNI     config: metrics section is empty, generating default config.go:370
00:41:04.540 WARNI     config: tracing section is empty, generating default config.go:370
00:41:04.541  INFO  consensus: cleaning empty Raft data folder (/home/nato/.ipfs-cluster/raft) raft.go:657
00:41:04.541 WARNI    service: the /home/nato/.ipfs-cluster/raft folder has been rotated.  Next start will use an empty state state.go:184
00:41:04.739  INFO    cluster: IPFS Cluster v0.10.0+git508131081124bc8289b762ffb7e0e0572798e9be listening on:
        /p2p-circuit/ipfs/QmTPuNCYQJTdsJ8xUZMGeZ9ReecMbsguaEqyk26s8TFUDV
        /ip4/127.0.0.1/tcp/9096/ipfs/QmTPuNCYQJTdsJ8xUZMGeZ9ReecMbsguaEqyk26s8TFUDV
        /ip4/192.168.1.2/tcp/9096/ipfs/QmTPuNCYQJTdsJ8xUZMGeZ9ReecMbsguaEqyk26s8TFUDV

 cluster.go:114
00:41:04.739  INFO  ipfsproxy: IPFS Proxy: /ip4/127.0.0.1/tcp/9095 -> /ip4/127.0.0.1/tcp/5001 ipfsproxy.go:273
00:41:04.739  INFO    restapi: REST API (HTTP): /ip4/127.0.0.1/tcp/9094 restapi.go:456
00:41:04.739  INFO  consensus: peer is ready to join a cluster raft.go:205
00:41:04.739  INFO    service: Bootstrapping to /ip4/159.89.116.13/tcp/9096/ipfs/QmUxeeoYLoxEQq37pPo2BiFPjcYvJLqeHBRMfL6YtWqtUA daemon.go:172
00:41:04.739  INFO    restapi: REST API (libp2p-http): ENABLED. Listening on:
        /p2p-circuit/ipfs/QmTPuNCYQJTdsJ8xUZMGeZ9ReecMbsguaEqyk26s8TFUDV
        /ip4/127.0.0.1/tcp/9096/ipfs/QmTPuNCYQJTdsJ8xUZMGeZ9ReecMbsguaEqyk26s8TFUDV
        /ip4/192.168.1.2/tcp/9096/ipfs/QmTPuNCYQJTdsJ8xUZMGeZ9ReecMbsguaEqyk26s8TFUDV

 restapi.go:473
00:41:05.239  INFO  consensus: Current Raft Leader: QmPp9hcP4ZzbkawU2nZ41kVQwHCZvdwenagQ6qzXWsJFnZ raft.go:280

00:41:05.264 ERROR   ipfshttp: error posting to IPFS: Post http://127.0.0.1:5001/api/v0/pin/ls?arg=QmZZExLhLfqhA4dpWs2iKB2w2cWQsuyYgSrv1Hnm7QBJXo&type=recursive: context canceled ipfshttp.go:431
00:41:05.264 ERROR  p2p-gorpc: Post http://127.0.0.1:5001/api/v0/pin/ls?arg=QmZZExLhLfqhA4dpWs2iKB2w2cWQsuyYgSrv1Hnm7QBJXo&type=recursive: context canceled call.go:63

00:41:06.050  INFO    cluster: Cluster Peers (without including ourselves): cluster.go:445
00:41:06.050  INFO    cluster:     - QmPp9hcP4ZzbkawU2nZ41kVQwHCZvdwenagQ6qzXWsJFnZ cluster.go:452
00:41:06.050  INFO    cluster:     - QmUxeeoYLoxEQq37pPo2BiFPjcYvJLqeHBRMfL6YtWqtUA cluster.go:452
00:41:06.050  INFO    cluster: ** IPFS Cluster is READY ** cluster.go:460

00:41:06.110 ERROR   ipfshttp: error posting to IPFS: Post http://127.0.0.1:5001/api/v0/pin/ls?arg=QmSbyFDGDrHtr2LKFX2fuSF3cChD54ZvphJKhfsh1SY7ZL&type=recursive: context canceled ipfshttp.go:431
00:41:06.110 ERROR  p2p-gorpc: Post http://127.0.0.1:5001/api/v0/pin/ls?arg=QmSbyFDGDrHtr2LKFX2fuSF3cChD54ZvphJKhfsh1SY7ZL&type=recursive: context canceled call.go:63

00:41:06.771  INFO  consensus: Current Raft Leader: QmPp9hcP4ZzbkawU2nZ41kVQwHCZvdwenagQ6qzXWsJFnZ raft.go:280
00:41:06.790  INFO    cluster: QmTPuNCYQJTdsJ8xUZMGeZ9ReecMbsguaEqyk26s8TFUDV: joined QmUxeeoYLoxEQq37pPo2BiFPjcYvJLqeHBRMfL6YtWqtUA's cluster cluster.go:760
00:41:11.080  INFO    cluster: QmTPuNCYQJTdsJ8xUZMGeZ9ReecMbsguaEqyk26s8TFUDV: removed from raft. Initiating shutdown cluster.go:357
00:41:11.080  INFO    cluster: shutting down Cluster cluster.go:483
00:41:11.080 ERROR  p2p-gorpc: context canceled call.go:63
00:41:11.080 ERROR  p2p-gorpc: context canceled call.go:63
00:41:11.080  INFO  consensus: stopping Consensus component consensus.go:185
00:41:12.604  INFO   ipfshttp: IPFS Unpin request succeeded: QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn ipfshttp.go:331
00:41:12.871 ERROR libp2p-raf: Failed to decode incoming command: transport shutdown transport.go:37
00:41:12.872  INFO  consensus: cleaning and backing up Raft data folder (/home/nato/.ipfs-cluster/raft) raft.go:662
00:41:12.873  INFO    monitor: stopping Monitor pubsubmon.go:162
00:41:12.873  INFO    restapi: stopping Cluster API restapi.go:494
00:41:12.873  INFO  ipfsproxy: stopping IPFS Proxy ipfsproxy.go:248
00:41:12.873  INFO   ipfshttp: stopping IPFS Connector ipfshttp.go:207
00:41:12.873  INFO pintracker: stopping MapPinTracker maptracker.go:124
00:41:12.873 ERROR  p2p-gorpc: context canceled call.go:63
00:41:12.874 ERROR  p2p-gorpc: context canceled call.go:63
00:41:12.874 ERROR  p2p-gorpc: context canceled call.go:63
```

## Problems

1. I can't join an existing cluster
2. I don't want to lose everything every time I'm to rejoin the cluster
3. I need leave-on-shutdown enabled so the cluster stays alive while a peer or two are disconnected

**Comment by hsanjuan** [comment]
hey, thanks for the detailed report.

This is a problem with Raft. Your new peer has probably not finished to apply all updates to the Raft state, so it checks and doesn't find itself as not being a member (probably looking at a moment after the log entry in which it was removed and before the log entry in which it was added again.

My impression was that this was fixed a while ago but it seems current safeguards don't work in all cases. The way you want to use Cluster will just be orders of magnitude easier once #685 lands. Until then, I can only offer you a work around consisting in increasing `peer_watch_interval` in the configuration to a larger value. I hope that works. The only other option would be to start your peer using a different peer ID/key.

Sorry for that..

**Comment by hsanjuan** [comment]
@NatoBoram now that #685 has landed I think I'm going to close this as I'm not sure we can do much about it. At least now we have a good alternative to clusters in which peers come and go.

**Comment by NatoBoram** [comment]
Alright, thanks! With fixes for #787 like 2d5b31ad6a94efcdcca33210cdf674f4939f7f2b, this shouldn't be a problem anymore.
